### PR TITLE
feat(abs): add backup sidecar logic

### PR DESCRIPTION
**Pull request by vdice**
Ref https://github.com/coreos/etcd-operator/pull/1237

### CODE REVIEW

File: pkg/backup/abs.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func (ab *absBackend) totalSize() (int64, error) {
+	return ab.ABS.TotalSize()
+}
+
+func (ab *absBackend) getBlobSize(key string) (int64, error) {
+	rc, err := ab.open(key)
+
+	b, err := ioutil.ReadAll(rc)
```
Comment (xiang90) [inline pkg/backup/abs.go]: this is pretty expensive :(

Comment (xiang90) [reply pkg/backup/abs.go]: maybe keep a smaller buf and discard the counted one. at least, add a todo.

**Comment by xiang90** [comment]
defer to @hongchaodeng and @hasbro17

File: pkg/backup/abs/absconfig/abs_ctx.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package absconfig
```
Comment (fanminshi) [inline pkg/backup/abs/absconfig/abs_ctx.go]: license title?

Comment (vdice) [reply pkg/backup/abs/absconfig/abs_ctx.go]: thanks; added.

**Comment by xiang90** [comment]
@hongchaodeng can we get this merged?

**Comment by hongchaodeng** [comment]
@etcd-bot test this please

File: pkg/backup/env/env.go [hunk 3]
```diff
@@ -15,7 +15,10 @@
...
 
 const (
-	ClusterSpec = "CLUSTER_SPEC"
-	AWSS3Bucket = "AWS_S3_BUCKET"
-	AWSConfig   = "AWS_CONFIG_FILE"
+	ClusterSpec       = "CLUSTER_SPEC"
+	AWSS3Bucket       = "AWS_S3_BUCKET"
+	AWSConfig         = "AWS_CONFIG_FILE"
+	ABSContainer      = "AZURE_STORAGE_CONTAINER"
+	ABSStorageAccount = "AZURE_STORAGE_ACCOUNT"
```
Comment (hongchaodeng) [inline pkg/backup/env/env.go]: This is not used

File: pkg/backup/abs/absconfig/abs_ctx.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import "errors"
+
+// ABSContext represents the controller context around a ABS backend configuration
+type ABSContext struct {
+	ABSSecret    string
+	ABSContainer string
+}
+
+// Validate validates a given ABSContext
+func (a *ABSContext) Validate() error {
```
Comment (hongchaodeng) [inline pkg/backup/abs/absconfig/abs_ctx.go]: This is not used in this PR

File: pkg/backup/abs/abs.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	return size, err
+}
+
+// CopyPrefix copies all blobs with given prefix
+func (w *ABS) CopyPrefix(from string) error {
+	_, blobs, err := w.list(from)
+	if err != nil {
+		return err
+	}
+	for _, blob := range blobs {
```
Comment (hongchaodeng) [inline pkg/backup/abs/abs.go]: blob -> key/basename

This is actually the basename of the path.

**Comment by vdice** [comment]
Feedback should now be addressed.  Just a quick note that I'll be away until Sep. 5th.  Thanks!

**Comment by hongchaodeng** [comment]
lgtm

Thanks @vdice

**Pull request merged**